File: VB/WindowsApplication3/CustomImageEdit.vb
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Repository
Imports System.Drawing
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Drawing
Imports System.ComponentModel
Imports DevExpress.XtraEditors.Controls

Namespace DXSample

    Public Class RepositoryItemCustomImageEdit
        Inherits RepositoryItemImageEdit

        Shared Sub New()
            Call RegisterCustomEdit()
        End Sub

        Public Sub New()
        End Sub

        Public Const CustomEditName As String = "CustomImageEdit"

        Public Overrides ReadOnly Property EditorTypeName As String
            Get
                Return CustomEditName
            End Get
        End Property

        Public Shared Sub RegisterCustomEdit()
            Call EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(CustomEditName, GetType(CustomImageEdit), GetType(RepositoryItemCustomImageEdit), GetType(ImageEditViewInfo), New CustomBlobBaseEditPainter(), True, CType(Nothing, System.Drawing.Image)))
        End Sub

        Public Shared Function GetImage(ByVal array As Byte()) As Image
            Return ByteImageConverter.FromByteArray(array)
        End Function
    End Class

    Public Class CustomImageEdit
        Inherits ImageEdit

        Shared Sub New()
            RepositoryItemCustomImageEdit.RegisterCustomEdit()
        End Sub

        Public Sub New()
        End Sub

        Public Overrides ReadOnly Property EditorTypeName As String
            Get
                Return RepositoryItemCustomImageEdit.CustomEditName
            End Get
        End Property

        <DesignerSerializationVisibility(DesignerSerializationVisibility.Content)>
        Public Overloads ReadOnly Property Properties As RepositoryItemCustomImageEdit
            Get
                Return TryCast(MyBase.Properties, RepositoryItemCustomImageEdit)
            End Get
        End Property
    End Class

    Public Class CustomBlobBaseEditPainter
        Inherits BlobBaseEditPainter

        Public Sub New()
            MyBase.New()
        End Sub

        Protected Overrides Sub DrawGlyphCore(ByVal info As ControlGraphicsInfoArgs, ByVal be As ButtonEditViewInfo)
            Dim info_ As ImageEditViewInfo = TryCast(be, ImageEditViewInfo)
            Dim image As Image = Nothing
            If TypeOf info_.EditValue Is Byte() Then
                image = New Bitmap(RepositoryItemCustomImageEdit.GetImage(TryCast(info_.EditValue, Byte())), info_.ImageSize)
            ElseIf TypeOf info_.EditValue Is Image Then
                image = New Bitmap(CType(info_.EditValue, Image), info_.ImageSize)
            End If

            If image Is Nothing Then
                MyBase.DrawGlyphCore(info, be)
                Return
            End If

            info.Cache.Paint.DrawImage(info.Graphics, image, be.GlyphBounds)
        End Sub
    End Class
End Namespace

File: CS/WindowsApplication3/CustomImageEdit.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using DevExpress.Skins;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Repository;
using System.Drawing;
using System.Reflection;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Drawing;
using System.IO;
using System.ComponentModel;
using DevExpress.XtraEditors.Controls;

namespace DXSample {
    public class RepositoryItemCustomImageEdit : RepositoryItemImageEdit
    {
        static RepositoryItemCustomImageEdit() { RegisterCustomEdit(); }

        public RepositoryItemCustomImageEdit() { }

        public const string CustomEditName = "CustomImageEdit";

        public override string EditorTypeName { get { return CustomEditName; } }

        public static void RegisterCustomEdit()
        {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(CustomEditName,
              typeof(CustomImageEdit), typeof(RepositoryItemCustomImageEdit),
              typeof(ImageEditViewInfo), new CustomBlobBaseEditPainter(), true, null));
        }

        public static Image GetImage(byte[] array)
        {
            return ByteImageConverter.FromByteArray(array);
        }
    }

    public class CustomImageEdit : ImageEdit {
        static CustomImageEdit() { RepositoryItemCustomImageEdit.RegisterCustomEdit(); }

        public CustomImageEdit() {  }

        public override string EditorTypeName
        {
            get {
                return RepositoryItemCustomImageEdit.CustomEditName;
            }
        }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemCustomImageEdit Properties
        {
            get { return base.Properties as RepositoryItemCustomImageEdit; }
        }
    }

    public class CustomBlobBaseEditPainter : BlobBaseEditPainter
    {
        public CustomBlobBaseEditPainter() : base() { }
        protected override void DrawGlyphCore(ControlGraphicsInfoArgs info, ButtonEditViewInfo be)
        {
            ImageEditViewInfo info_ = be as ImageEditViewInfo;
            Image image = null;
            if (info_.EditValue is byte[])
                image = new Bitmap(RepositoryItemCustomImageEdit.GetImage(info_.EditValue as byte[]), info_.ImageSize);
            else if (info_.EditValue is Image)
                image = new Bitmap((Image)info_.EditValue, info_.ImageSize);
            if (image == null)
            {
                base.DrawGlyphCore(info, be);
                return;
            }
            info.Cache.Paint.DrawImage(info.Graphics, image, be.GlyphBounds);
        }
    }
}